### PR TITLE
Reviews tab: refactor to have one instance per site ID

### PR DESCRIPTION
**Pull request by jaclync**
Part of #1838 

## Background

In this PR, a new observable for the currently logged in site ID is introduced to `SessionManager`/`StorageManager` so that the main tab (`MainTabBarController`) can subscribe to site ID changes.

- When the app is logged out (`nil` site ID), the main tab's view controllers become empty
- If the user is logged in to a site (non-nil site ID), the main tab's view controllers are set up given the site ID (only Reviews tab is taking in the site ID in this PR; other tabs will follow in future PRs)

Any feedback on the observable implementation is welcome!

## Changes

- Added an observable site ID to `SessionManager` and `StoresManager` (just forwarding `SessionManager`'s observable)
  - Since the site ID is stored in `UserDefaults`, I created `UserDefaults.defaultStoreID` so that we can observe this dynamic var. The return type has to be Objective-C compatible for KVO, so I converted `Int` to `Int64` before emitting the value.
  - Updated the struct to a class, so that we can invalidate the KVO (even though it's a singleton in the app). This also allowed us to update some `var` to `let` in unit tests.
- In `MainTabBarController`'s `viewDidLoad`, replaced the current `viewControllers` setup with subscription for site ID changes so that `viewControllers` follow the behavior:
  - Nil site ID: `viewControllers` are empty, and the site ID specific Reviews coordinator is set to nil
  - Non-nil site ID: `viewControllers` are set up with the existing Dashboard, Orders, and Products tabs plus a new instance of Reviews coordinator that is tied to a site ID. In future PRs, Dashboard, Orders, and Products tabs are also going to be torn down on each site ID change.
- DI'ed site ID to Reviews tab classes - `DefaultReviewsDataSource`, `ReviewDetailsViewController` `ReviewsCoordinator`, `ReviewsViewController`, `ReviewsViewModel`. In `ReviewsViewController`, removed subscription for `StoresManagerDidUpdateDefaultSite` notifications to reload results controller and resync data.
- Added unit tests for `DefaultStoresManager` and `MainTabBarController`

### Logging in / switching stores

- Log out from the app
- Log in to an account with multiple sites, select a site to continue --> all the tabs should load as before
- Navigate to the Reviews tab --> the reviews should be loaded for the selected site
- Navigate to the My Store tab, tap Settings, then switch to the same store --> nothing should happen
- Tap Settings, then switch to a different store --> all the tabs should be updated to load data for the selected site

### Review push notifications

- Enable dev push notifications on sandbox p99K0U-1qz-p2
- Log in and navigate to a tab that is not Reviews
- Place the app in the background
- Submit a product review on the web
- Wait for the push notification
- Tap on the push notification
- Wait for a second --> the app should navigate to the Reviews tab and show the review details

Update release notes:
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/Classes/ViewRelated/MainTabBarController.swift
Comment (Ecarrion) [inline WooCommerce/Classes/ViewRelated/MainTabBarController.swift]: One thing I don't quite understand well is why we set up `viewControllers` to an empty array but still hold a reference to all root view controllers. Is this something that will change once you refactor the other VC's to accept a site ID?

File: WooCommerce/Classes/System/SessionManager.swift
Comment (Ecarrion) [inline WooCommerce/Classes/System/SessionManager.swift:26]: Not sure if possible, but can't we expose this as an observable so we don't have to deal with KVO observations? 😅

**Review by Ecarrion — COMMENTED** [review]
Code looks good! I just have some questions to understand it correctly.
Testing also looks, good but I think I found a minor glitch(VC starts popping but then it changes suddenly) on the animation while logging out.
I think we didn't have any animation before but the glitch does look a bit weird 😅 

logout

File: WooCommerce/Classes/ViewRelated/MainTabBarController.swift
Comment (jaclync) [reply WooCommerce/Classes/ViewRelated/MainTabBarController.swift]: This PR only has the Reviews tab associated with a site ID, and the other tabs still rely on listening to the site change notification so I keep them around. After we finalize the approach for the Reviews tab, I'll do the same for other tab view controllers so that all of them will be set to `nil` here.

File: WooCommerce/Classes/System/SessionManager.swift
Comment (jaclync) [reply WooCommerce/Classes/System/SessionManager.swift:26]: Did you mean opening this up directly for whoever interested in the default store ID to subscribe? `UserDefaults` seems an implementation detail of `SessionManager`, and that's why I created a public observable `siteID: Observable<Int64?>` from `UserDefaults` KVO. Also, `UserDefaults` KVO has some limitation since it's ObjC API and doesn't work with `Int64` directly.

Not sure if I misunderstood anything, feel free to suggest another way!

**Comment by jaclync** [comment]
> Testing also looks, good but I think I found a minor glitch(VC starts popping but then it changes suddenly) on the animation while logging out.
I think we didn't have any animation before but the glitch does look a bit weird 😅

Nice catch @Ecarrion! When the user logs out, we deauthenticate the app, then we pop the dashboard view controller's navigation controller to root animated:

https://github.com/woocommerce/woocommerce-ios/blob/ed8be1f7e79e975545fa384b64d832719d83495b/WooCommerce/Classes/ViewRelated/Dashboard/Settings/SettingsViewController.swift#L379-L382

`ServiceLocator.stores.deauthenticate()` triggers the `state` to change from `AuthenticatedState` to `DeauthenticatedState`, and `DeauthenticatedState`'s `didEnter` event triggers the app delegate to display authentication UI:

https://github.com/woocommerce/woocommerce-ios/blob/ed8be1f7e79e975545fa384b64d832719d83495b/WooCommerce/Classes/Yosemite/DeauthenticatedState.swift#L10-L14

In the meantime, this PR also empties the main tab bar's `viewControllers`. I'm guessing this interferes with the `navigationController?.popToRootViewController(animated: true)` in `SettingsViewController`'s log out above and causes the weird animation.

In https://github.com/woocommerce/woocommerce-ios/pull/2850/commits/3436ef3b61c44984c4cf77c6cd1feaa4877f8bd2, I removed the line to pop to root view controller animated and then turn on animation for displaying the authentication UI:

simulator

@Garance91540 does the animation after logging out look good to you? ☝️ 

This is what it looks like without animating the authentication UI:

simulator

**Comment by Garance91540** [comment]
> @Garance91540 does the animation after logging out look good to you? ☝️

Yes! definitily add some smooth in there :)

File: WooCommerce/Classes/System/SessionManager.swift
Comment (Ecarrion) [reply WooCommerce/Classes/System/SessionManager.swift:26]: NVM, I think I misunderstood this 😇

**Comment by Ecarrion** [comment]
@jackiejade I just test this and while the animation looks great I think we introduced another bug, but I'm not sure if it is by [not clearing the navigation controller](https://github.com/woocommerce/woocommerce-ios/pull/2850/commits/3436ef3b61c44984c4cf77c6cd1feaa4877f8bd2#diff-56863e4b60e22cc16f786c5393dec589L381).

- Log in to a store in the app
- Navigate to orders and tap on one
- Log out
- Log in to a different store
- Navigate to orders and see that the order from the previous store is still visible

**Comment by jaclync** [comment]
> I just test this and while the animation looks great I think we introduced another bug, but I'm not sure if it is by not clearing the navigation controller.
Log in to a store in the app
Navigate to orders and tap on one
Log out
Log in to a different store
Navigate to orders and see that the order from the previous store is still visible

@Ecarrion I think you mentioned another Jackie 😄 Thanks for testing and noticing this one, I was able to reproduce this issue in production as well. Also, this issue occurs after switching to another store without logging out as well. The code to [not pop the navigation controller to root](https://github.com/woocommerce/woocommerce-ios/pull/2850/commits/3436ef3b61c44984c4cf77c6cd1feaa4877f8bd2#diff-56863e4b60e22cc16f786c5393dec589L381) just affects the dashboard tab (settings VC's navigation stack).

I just updated the PR with this commit https://github.com/woocommerce/woocommerce-ios/pull/2850/commits/690c59e08fdca0b38dd4440cbb81ae83012ae852 to recreate all tab view controllers, without DI'ing the site ID to Dashboard/Orders/Products tabs. This should fix the issue above, and all tab view controllers now have life cycle tied to a logged in site. In the future, we can update the Dashboard/Orders/Products view controllers to DI site ID and remove their default site change notification observation. Lemme know what you think!

**Review by Ecarrion — APPROVED** [review]
👏 
Thanks, @jaclync for the changes. I tested the logout and switch store flow and it's working correctly ✅ 

The only thing I noticed was a strange notice message when changing store "700 products have been deleted from your store" which I assume it's because we are clearing the DB from the old store.  

Should I file a bug for that scenario or is that expected?

**Comment by Ecarrion** [comment]
Also sorry for mentioning the wrong person 🤦

File: WooCommerce/Classes/ViewRelated/MainTabBarController.swift
Comment (shiki) [inline WooCommerce/Classes/ViewRelated/MainTabBarController.swift:87]: It's probably best to still add `cancellableSiteID.cancel()` on `deinit` even though `MainTabBarController` lives forever.

File: WooCommerce/Classes/ViewRelated/MainTabBarController.swift
Comment (shiki) [inline WooCommerce/Classes/ViewRelated/MainTabBarController.swift]: These properties are not used anywhere outside of `updateViewControllers`. It's probably best to remove them unless we're planning to use these later. 

```suggestion

```

The `reviewsTabCoordinator` is needed though. It needs to be kept in memory because it uses an observer too. Bummer.

File: WooCommerce/Classes/ViewRelated/Reviews/ReviewDetailsViewController.swift
Comment (shiki) [inline WooCommerce/Classes/ViewRelated/Reviews/ReviewDetailsViewController.swift]: I _think_ we can infer the `siteID` from `ProductReview.siteID`. We can keep the `siteID` property but we don't need to have the `init `argument. What do you think? 🙂

File: WooCommerce/Classes/ViewRelated/Reviews/ReviewDetailsViewController.swift
Comment (shiki) [inline WooCommerce/Classes/ViewRelated/Reviews/ReviewDetailsViewController.swift:167]: Fantastic! 🤸

File: WooCommerce/Classes/ViewRelated/Reviews/ReviewsViewController.swift
Comment (shiki) [inline WooCommerce/Classes/ViewRelated/Reviews/ReviewsViewController.swift:364]: 👏

File: WooCommerce/WooCommerceTests/Internal/SessionManager+Internal.swift
Comment (shiki) [inline WooCommerce/WooCommerceTests/Internal/SessionManager+Internal.swift]: ⚠️ We should create an instance now instead of using `testingInstance`. The effect of changing `SessionManager` from `struct` to `class` is that `SessionManager.testingInstance` is now **a singleton**. 🙈 That can become problematic since its `defaultCredentials` property is mutable. 

```suggestion
        let manager = SessionManager(defaults: SessionSettings.defaults, keychainServiceName: SessionSettings.keychainServiceName)
```

We should get rid of `testingInstance` later and test cases should use `makeForTesting()`.

File: WooCommerce/WooCommerceTests/ViewRelated/MainTabBarControllerTests.swift
Comment (shiki) [inline WooCommerce/WooCommerceTests/ViewRelated/MainTabBarControllerTests.swift:14]: The disadvantage of `ServiceLocator`. I can't tell when or where these are used. 🤪

File: WooCommerce/WooCommerceTests/ViewRelated/MainTabBarControllerTests.swift
Comment (shiki) [inline WooCommerce/WooCommerceTests/ViewRelated/MainTabBarControllerTests.swift]: Maybe: 

```suggestion
    func test_viewControllers_are_replaced_after_updating_to_a_different_site() throws {
```

Or something along those lines. Just wanted to highlight that it's not just “reviews” tab anymore. 😅

File: WooCommerce/WooCommerceTests/ViewRelated/MainTabBarControllerTests.swift
Comment (shiki) [inline WooCommerce/WooCommerceTests/ViewRelated/MainTabBarControllerTests.swift]: What do you think about adding an assertion for the previous state? 

```suggestion
        let siteID: Int64 = 134
        stores.updateDefaultStore(storeID: siteID)

        XCTAssertEqual(tabBarController.viewControllers?.count, 4)

        // Action
```

Maybe it will help with understanding more what changed.

File: WooCommerce/WooCommerceTests/ViewRelated/MainTabBarControllerTests.swift
Comment (shiki) [inline WooCommerce/WooCommerceTests/ViewRelated/MainTabBarControllerTests.swift:5]: Thanks for adding tests! This is awesome! 👏

**Review by shiki — COMMENTED** [review]
This is great! Thank you for working on this! 

✔️ Tested store switching
✔️ Tested logging out
✔️ Checked memory after logging out
✔️ Tested order and review push notifications

I just have a few suggestions and some celebratory comments. 😁 

## Objects Not Deallocated

These are separate problems and I'm glad that this PR helps us with catching these.

Looks like `DashboardViewController` is not deallocated after logging out. 

image

Also, and this is terrible. And obviously my mistake. I forgot to add a `weak self` to the `Publisher.sink` calls. So `OrderListViewController` instances are also not deallocated. 😅 

image

Would you mind creating issues for this after this PR is merged? I can tackle the OrderList one. 😄

File: WooCommerce/Classes/ViewRelated/MainTabBarController.swift
Comment (jaclync) [reply WooCommerce/Classes/ViewRelated/MainTabBarController.swift:87]: yea sounds good! updated in https://github.com/woocommerce/woocommerce-ios/pull/2850/commits/5a9d65197d94f1343205132c6b5f2d20adbf6c17

File: WooCommerce/Classes/ViewRelated/MainTabBarController.swift
Comment (jaclync) [reply WooCommerce/Classes/ViewRelated/MainTabBarController.swift]: yup `reviewsTabCoordinator` is needed for push notification presentation! removed the rest in https://github.com/woocommerce/woocommerce-ios/pull/2850/commits/5a9d65197d94f1343205132c6b5f2d20adbf6c17

File: WooCommerce/Classes/ViewRelated/Reviews/ReviewDetailsViewController.swift
Comment (jaclync) [reply WooCommerce/Classes/ViewRelated/Reviews/ReviewDetailsViewController.swift]: makes sense! updated in https://github.com/woocommerce/woocommerce-ios/pull/2850/commits/d8f33ecce5c04540fc3f7cb16f6be691e1013920

File: WooCommerce/WooCommerceTests/ViewRelated/MainTabBarControllerTests.swift
Comment (jaclync) [reply WooCommerce/WooCommerceTests/ViewRelated/MainTabBarControllerTests.swift]: oops I forgot to update the test case name after replacing all tab view controllers! updated in https://github.com/woocommerce/woocommerce-ios/pull/2850/commits/d8f33ecce5c04540fc3f7cb16f6be691e1013920

File: WooCommerce/WooCommerceTests/ViewRelated/MainTabBarControllerTests.swift
Comment (jaclync) [reply WooCommerce/WooCommerceTests/ViewRelated/MainTabBarControllerTests.swift]: yea good suggestion! updated in https://github.com/woocommerce/woocommerce-ios/pull/2850/commits/d8f33ecce5c04540fc3f7cb16f6be691e1013920

File: WooCommerce/WooCommerceTests/Internal/SessionManager+Internal.swift
Comment (jaclync) [reply WooCommerce/WooCommerceTests/Internal/SessionManager+Internal.swift]: ooh I actually thought it generates a new instance of in `.testingInstance` until I saw the `static`! I updated `makeForTesting()` to generate a new instance in https://github.com/woocommerce/woocommerce-ios/pull/2850/commits/d8f33ecce5c04540fc3f7cb16f6be691e1013920, and created an issue for the replacement in https://github.com/woocommerce/woocommerce-ios/issues/2885

File: WooCommerce/WooCommerceTests/ViewRelated/MainTabBarControllerTests.swift
Comment (jaclync) [reply WooCommerce/WooCommerceTests/ViewRelated/MainTabBarControllerTests.swift:14]: yea, I'm not a fan of this 😞 however, `MainTabBarController` is initialized from `Main.storyboard` right now. It'd be nice to remove `Main.storyboard` at some point so we can DI to `MainTabBarController` more easily.

**Comment by jaclync** [comment]
@shiki thanks for the review and noticing these retain issues after logout! I updated the PR and responded to your comments. Regarding the retain issues after logout:

> Looks like DashboardViewController is not deallocated after logging out.

This is very interesting! I took a look and I believe it's `MainTabBarController`'s `selectedViewController` holding onto `DashboardViewController` and I couldn't easily reset `selectedViewController`. I tried a few things and will fix this in a separate issue: https://github.com/woocommerce/woocommerce-ios/issues/2884

> Also, and this is terrible. And obviously my mistake. I forgot to add a weak self to the Publisher.sink calls. So OrderListViewController instances are also not deallocated. 😅

This is an easy fix, exactly what you said about the sink call! I fixed this in https://github.com/woocommerce/woocommerce-ios/pull/2850/commits/6d1ea95434499db18cc2ac7ea56efe0423944a18 if you don't mind 🙂

**Comment by jaclync** [comment]
> The only thing I noticed was a strange notice message when changing store "700 products have been deleted from your store" which I assume it's because we are clearing the DB from the old store.
Should I file a bug for that scenario or is that expected?

@Ecarrion somehow I haven't seen the notice message after switching stores 🤔 I see the notice about site switch:



Do you happen to have a screenshot or steps to reproduce this "deleted" notice message?

**Comment by jaclync** [comment]
I created another PR https://github.com/woocommerce/woocommerce-ios/pull/2886 to fix the `DashboardViewController` being retained after logging out. The PR updates the implementation in `MainTabBarController` a bit with authentication animation improvement (only animates when transitioning from logged in to logged out state). Since this PR has been reviewed and contains other changes like the site ID in Reviews tab classes, I'm planning to merge both PRs in the same release when they are ready. Lemme know if you have other suggestions!

**Comment by Ecarrion** [comment]
> Do you happen to have a screenshot or steps to reproduce this "deleted" notice message?

I think I misread the message it says: "order xxx has been deleted from your store" 😛 

And it happens if you do the following
- Open the app
- Navigate to orders tab
- Open an order
- Tap on "my store" tab
- Change store
- See notice message

Screen Shot 2020-09-29 at 7 34 50 AM

**Comment by shiki** [comment]
@Ecarrion That deletion message looks like https://github.com/woocommerce/woocommerce-ios/issues/2369. It's quite misleading and we should remove it soon. 🙈

**Review by shiki — APPROVED** [review]
👏

**Comment by jaclync** [comment]
> the message it says: "order xxx has been deleted from your store" 😛

thanks for the steps to repro Ernesto and pointing to the existing issue #2369 Jayson! yea I agree it makes sense to remove the notice message, since we delete all the stored orders when the user pulls down to refresh on the orders tab as well and we currently don't have UI to manually delete an order.

merging now to unblock other open PRs!

**Pull request merged**
